operations/image_management/Upload_and_Register_an_Image_Recipe.md

```diff
@@ -131,3 +131,145 @@ Download and expand recipe archives from S3 and IMS. Modify and upload a recipe
     type = "s3"
     ```
 
+## Templating IMS recipes when building an IMS image
+
+IMS can optionally template the contents of an IMS recipe before building the recipe with kiwi-ng during an IMS create job. This 
+enables variables stored in the recipe to be dynamically replaced with values registered with the IMS recipe record after the 
+recipe is downloaded from S3 but before it is built by Kiwi-ng. This functionality is used by various Cray products to ensure 
+that the resulting image is built from the most correct release versions of Nexus hosted RPM repositories. 
+
+Follow the steps below to enable IMS templating within a given recipe. Note that for IMS to properly template a recipe,
+the following procedures must all be completed.
+
+### Enable templating in IMS recipe archive tgz file
+
+1. Add a file named '.ims_recipe_template.yaml' to the root of the recipe archive. The contents of this file must be valid
+   yaml containing the keyword `template_files`. The `template_files` keyword must contain a list of files to be templated
+   within the image recipe archive. 
+
+   ``` bash
+   ncn# ls -al
+   drwxr-xr-x 3 root root    70 Feb 22 19:05 .
+   drwxr-xr-x 3 root root    41 Feb 22 17:18 ..      
+   -rw-r--r-- 1 root root  2744 Feb  4 21:14 config.sh
+   -rw-r--r-- 1 root root 24645 Feb  4 21:15 config.xml
+   -rw-r--r-- 1 root root  1216 Feb  4 21:14 images.sh
+   -rw-r--r-- 1 root root    29 Feb  4 21:14 .ims_recipe_template.yaml
+   drwxr-xr-x 3 root root    18 Feb  4 21:15 root
+      
+   ncn# cat .ims_recipe_template.yaml
+   template_files:
+   - config.xml
+   ```
+   
+   IMS will template the indicated files 'in place' -- that is, the listed file will be replaced with the templated copy.
+   NOTE: Only files listed under the template_files key will be templated by IMS. 
+
+2. For each indicated file above, add appropriate Jinja2 variables where required to enable the result that are sought.
+   In the example below, the Jinja2 variable `{{ CSM_RELEASE_VERSION }}` is used to help identify the fully qualified 
+   Nexus repo name/path to use when building the recipe.
+
+   ```bash
+   ncn# cat config.xml
+   ...
+   <repository type="rpm-md" alias="csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3" priority="2" imageinclude="true">
+       <source path="https://packages.local/repository/csm-{{ CSM_RELEASE_VERSION }}-sle-15sp3/"/>
+   </repository>
+   ...
+   ```
+
+   NOTE: The repo referenced above is for documentation purposes only and may not actually exist. 
+
+3. Create a tgz archive of the image recipe.
+
+    ```bash
+    ncn# export ARTIFACT_FILENAME=recipe.tar.gz
+    ncn# tar cvfz ../$ARTIFACT_FILE .
+    ncn# cd ..
+    ```
+
+### Add template key/value pairs to IMS recipe record
```

Comment:
Change to:

Add Template Key/Value Pairs to IMS Recipe Record

Reply:
Done